src/librustc_traits/lowering.rs

```diff
@@ -99,6 +101,38 @@ impl<'tcx> Lower<Goal<'tcx>> for ty::Predicate<'tcx> {
     }
 }
 
+/// For a given `where_clause`, returns a clause `FromEnv(WC) :- Implemented(Self: Trait<P1..Pn>)`.
```

Comment:
So actually the rule we want is `FromEnv(WC) :- FromEnv(Self: Trait<P1..Pn>)`. But there is a slight subtlety: the `FromEnv` in the left hand side is not exactly a real `FromEnv` predicate, it's rather some notation (maybe not very appropriate, I agree):
* if `WC == TraitRef`, then indeed it means `FromEnv(TraitRef)`
* if `WC == ProjectionPredicate`, then again it means `FromEnv(ProjectionPredicate)`
* in any other case, `WC` is left unchanged, e.g. in the above `FromEnv(T: 'a)` just means `T: 'a`
see [this section in the book](https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html#transforming-the-lowered-where-clauses).

As an example, given this trait:
```rust
trait Foo<'a, T> where T: Copy, T: 'a, T: Iterator<Item = i32> { }
```
we would like the following clauses:
```
forall<'a, Self, T> {
    FromEnv(T: Copy) :- FromEnv(Self: Foo<'a, T>).
    FromEnv(T: Iterator<Item = i32>) :- FromEnv(Self: Foo<'a, T>).
    (T: 'a) :- FromEnv(Self: Foo<'a, T>).
}
```

Reply:
Oops, I really don't know how I mixed that up >.>

Doing the rest in a match clause really ought to be a matter of using `Lower` IMO, but right now it eagerly lowers straight to a `Goal` when `DomainGoal` is needed. I want to play around with this when I get some time.

Reply:
Yeah, I think I'm going to issue a fix for the eager lowering soon, it should benefit the ergonomics of your PR